File: apprefurbished/src/main/java/com/androidcollider/easyfin/transactions/add_edit/income_expense/AddTransactionIncomeExpenseModel.java
package com.androidcollider.easyfin.transactions.add_edit.income_expense;

import androidx.core.util.Pair;

import com.androidcollider.easyfin.common.managers.accounts.accounts_to_spin_view_model.AccountsToSpinViewModelManager;
import com.androidcollider.easyfin.common.managers.format.date.DateFormatManager;
import com.androidcollider.easyfin.common.managers.format.number.NumberFormatManager;
import com.androidcollider.easyfin.common.models.Transaction;
import com.androidcollider.easyfin.common.models.TransactionCategory;
import com.androidcollider.easyfin.common.repository.Repository;
import com.androidcollider.easyfin.common.view_models.SpinAccountViewModel;

import java.util.List;

import io.reactivex.rxjava3.core.Flowable;

/**
 * @author Ihor Bilous
 */

class AddTransactionIncomeExpenseModel implements AddTransactionIncomeExpenseMVP.Model {

    private Repository repository;
    private NumberFormatManager numberFormatManager;
    private DateFormatManager dateFormatManager;
    private AccountsToSpinViewModelManager accountsToSpinViewModelManager;


    AddTransactionIncomeExpenseModel(Repository repository,
                                     NumberFormatManager numberFormatManager,
                                     DateFormatManager dateFormatManager,
                                     AccountsToSpinViewModelManager accountsToSpinViewModelManager) {
        this.repository = repository;
        this.numberFormatManager = numberFormatManager;
        this.dateFormatManager = dateFormatManager;
        this.accountsToSpinViewModelManager = accountsToSpinViewModelManager;
    }

    @Override
    public Flowable<Pair<List<SpinAccountViewModel>, List<TransactionCategory>>> getAccountsAndTransactionCategories(boolean isExpense) {
        return Flowable.combineLatest(
                accountsToSpinViewModelManager.getSpinAccountViewModelList(repository.getAllAccounts()),
                getTransactionCategories(isExpense),
                Pair::new
        );
    }

    @Override
    public Flowable<List<TransactionCategory>> getTransactionCategories(boolean isExpense) {
        return isExpense ?
                repository.getAllTransactionExpenseCategories() :
                repository.getAllTransactionIncomeCategories();
    }

    @Override
    public Flowable<Transaction> addNewTransaction(Transaction transaction) {
        return repository.addNewTransaction(transaction);
    }

    @Override
    public Flowable<Transaction> updateTransaction(Transaction transaction) {
        return repository.updateTransaction(transaction);
    }

    @Override
    public Flowable<Boolean> updateTransactionDifferentAccounts(Transaction transaction, double oldAccountAmount, int oldAccountId) {
        return repository.updateTransactionDifferentAccounts(transaction, oldAccountAmount, oldAccountId);
    }

    @Override
    public Flowable<TransactionCategory> addNewTransactionCategory(TransactionCategory transactionCategory, boolean isExpense) {
        return isExpense ?
                repository.addNewTransactionExpenseCategory(transactionCategory) :
                repository.addNewTransactionIncomeCategory(transactionCategory);
    }

    @Override
    public String prepareStringToParse(String value) {
        return numberFormatManager.prepareStringToParse(value);
    }

    @Override
    public long getMillisFromString(String date) {
        return dateFormatManager.stringToDate(date, DateFormatManager.DAY_MONTH_YEAR_SPACED).getTime();
    }

    @Override
    public boolean isDoubleNegative(double d) {
        return numberFormatManager.isDoubleNegative(d);
    }

    @Override
    public String getTransactionForEditAmount(int type, double amount) {
        return numberFormatManager.doubleToStringFormatterForEdit(
                type == 1 ? amount : Math.abs(amount),
                NumberFormatManager.FORMAT_1,
                NumberFormatManager.PRECISE_1
        );
    }
}

File: apprefurbished/src/main/java/com/androidcollider/easyfin/transactions/add_edit/income_expense/AddTransactionIncomeExpenseModel.kt
package com.androidcollider.easyfin.transactions.add_edit.income_expense

import com.androidcollider.easyfin.common.managers.accounts.accounts_to_spin_view_model.AccountsToSpinViewModelManager
import com.androidcollider.easyfin.common.managers.format.date.DateFormatManager
import com.androidcollider.easyfin.common.managers.format.number.NumberFormatManager
import com.androidcollider.easyfin.common.models.Transaction
import com.androidcollider.easyfin.common.models.TransactionCategory
import com.androidcollider.easyfin.common.repository.Repository
import com.androidcollider.easyfin.common.view_models.SpinAccountViewModel
import io.reactivex.rxjava3.core.Single
import kotlin.math.abs

/**
 * @author Ihor Bilous
 */
internal class AddTransactionIncomeExpenseModel(
    private val repository: Repository,
    private val numberFormatManager: NumberFormatManager,
    private val dateFormatManager: DateFormatManager,
    private val accountsToSpinViewModelManager: AccountsToSpinViewModelManager
) : AddTransactionIncomeExpenseMVP.Model {

    override fun getAccountsAndTransactionCategories(isExpense: Boolean):
            Single<Pair<List<SpinAccountViewModel>, List<TransactionCategory>>> {
        return Single.zip(
            accountsToSpinViewModelManager.getSpinAccountViewModelList(repository.allAccounts!!),
            getTransactionCategories(isExpense)
        ) { first: List<SpinAccountViewModel>,
            second: List<TransactionCategory> ->
            Pair(first, second)
        }
    }

    override fun getTransactionCategories(isExpense: Boolean): Single<List<TransactionCategory>> {
        return if (isExpense) repository.allTransactionExpenseCategories!!
        else repository.allTransactionIncomeCategories!!
    }

    override fun addNewTransaction(transaction: Transaction): Single<Transaction> {
        return repository.addNewTransaction(transaction)
    }

    override fun updateTransaction(transaction: Transaction): Single<Transaction> {
        return repository.updateTransaction(transaction)
    }

    override fun updateTransactionDifferentAccounts(
        transaction: Transaction,
        oldAccountAmount: Double,
        oldAccountId: Int
    ): Single<Boolean> {
        return repository.updateTransactionDifferentAccounts(
            transaction,
            oldAccountAmount,
            oldAccountId
        )
    }

    override fun addNewTransactionCategory(
        transactionCategory: TransactionCategory,
        isExpense: Boolean
    ): Single<TransactionCategory> {
        return if (isExpense) repository.addNewTransactionExpenseCategory(transactionCategory)
        else repository.addNewTransactionIncomeCategory(transactionCategory)
    }

    override fun prepareStringToParse(value: String): String {
        return numberFormatManager.prepareStringToParse(value)
    }

    override fun getMillisFromString(date: String): Long {
        return dateFormatManager.stringToDate(date, DateFormatManager.DAY_MONTH_YEAR_SPACED).time
    }

    override fun isDoubleNegative(d: Double): Boolean {
        return numberFormatManager.isDoubleNegative(d)
    }

    override fun getTransactionForEditAmount(type: Int, amount: Double): String {
        return numberFormatManager.doubleToStringFormatterForEdit(
            if (type == 1) amount else abs(amount),
            NumberFormatManager.FORMAT_1,
            NumberFormatManager.PRECISE_1
        )
    }
}